Sources/PackageExtension/PackageExtension.swift

```diff
@@ -0,0 +1,339 @@
+    /// SwiftPM version).
+    public let moduleName: String
+    
+    /// The path of the target directory.
+    public let targetDir: Path
+    
+    /// That path of the package that contains the target.
+    public let packageDir: Path
+    
+    /// Absolute paths of the source files in the target. This might include
+    /// derived source files generated by other extensions).
+    public let sourceFiles: [Path]
+
+    /// Absolute paths of the resource files in the target.
+    public let resourceFiles: [Path]
+
+    /// Absolute paths of any other files (not sources or resources) in the target.
+    public let otherFiles: [Path]
+
+    /// Information about all targets on which the target for which the exten-
+    /// sion is being invoked either directly or indirectly depends. This list
+    /// is in topologically sorted order, with immediate dependencies appearing
+    /// earlier and more distant dependencies later in the list. This is mainly
+    /// intended for generating lists of search path arguments, etc.
+    public let dependencies: [DependencyTargetInfo]
+    
+    /// Provides information about a target on which the target being built depends.
+    public struct DependencyTargetInfo: Decodable {
+        
+        /// The name of the target.
+        public let targetName: String
+        
+        /// The module name of the target. This is usually derived from the name,
+        /// but is possibly be customizable in the package manifest in some future
+        /// SwiftPM version).
+        public let moduleName: String
+        
+        /// The path of the target directory.
+        public let targetDir: Path
+    }
+
+    /// The path of an output directory into which files generated by the build
+    /// commands that are set up by the package extension can be written.  The
+    /// package extension itself may also write to this directory.
+    public let outputDir: Path
+    
+    /// A directory into which the package extension or the tool it invokes can
+    /// write any caches that speed up its operation or any other intermediate
+    /// files that shouldn't be further processed.
+    public let cacheDir: Path
+    
+    /// Looks up and returns the path of a named command line executable tool.
+    /// The executable must be either in the toolchain or in the system search
+    /// path for executables, or be provided by an executable target or binary
+    /// target on which the package extension target depends. Returns nil, but
+    /// does not throw an error, if the tool isn't found. Extensions that re-
+    /// quire the tool should emit an error diagnostic if it cannot be found.
+    public func lookupTool(named name: String) throws -> Path {
+        // TODO: Rather than just appending the name, this should instead use
+        // a mapping of tool names to paths (passed in from the context).
+        return execsDir.appending(name)
```

Comment:
This is preliminary.  See the comments.

Reply:
right. makes sense for Context to take in a function that will do the lookup, or a name:path map

Reply:
Absolutely.  In particular, it should be an error to depend on anything you didn't declare a dependency on.

My only concern is that this will preclude quick workarounds, such as dropping a tool into /usr/local/bin etc.  Perhaps that should be a warning if you use a tool you didn't depend on.

One could also argue that a standard set of tools in the toolchain should be fair game, e.g. `strip` etc.  Certainly `awk` and others seem reasonable.  But it's something to be discussed, I think.

Reply:
With all the things in this PR, I'm trying to stub out things that can be decided later if we think that it will "slot into" the current approach in a reasonable way.  This is one of those where I think a name-to-path map would be reasonable.